{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "!pip install openpyxl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import openpyxl\n",
    "from rdflib import Namespace, Graph, URIRef, Literal, BNode\n",
    "from rdflib.namespace import DCTERMS, DC, XSD, RDFS, DCAT, RDF, FOAF\n",
    "import os\n",
    "import urllib.parse\n",
    "from datetime import datetime\n",
    "\n",
    "g = Graph()\n",
    "WD = Namespace(\"http://www.wikidata.org/entity/\")\n",
    "WDT = Namespace(\"http://www.wikidata.org/prop/direct/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Graph identifier=Nc26dc543f80b4b00aecdb6329e712d1d (<class 'rdflib.graph.Graph'>)>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fileURI = URIRef(\"https://g-c8521c.81de.36fe.data.globus.org/Bayraktar/dt_Bayraktar_ClinicalMetadata_mock_2022.07.21.xlsx\")\n",
    "\n",
    "g.add((fileURI, RDF.type, DCAT.Dataset))\n",
    "g.add((fileURI, DCTERMS.format, WD.Q63082925))\n",
    "filename = 'dt_Bayraktar_ClinicalMetadata_mock_2022.07.21.xlsx'\n",
    "g.add((fileURI, RDFS.label, Literal(filename, datatype=XSD.string)))\n",
    "datecreated = os.path.getctime(filename)\n",
    "g.add((fileURI, DCTERMS.issued, Literal(datetime.fromtimestamp(datecreated).strftime('%Y-%m-%dT%H:%M:%S'), datatype=XSD.date)))\n",
    "datemodified = os.path.getmtime(filename)\n",
    "g.add((fileURI, DCTERMS.modified, Literal(datetime.fromtimestamp(datemodified).strftime('%Y-%m-%dT%H:%M:%S'), datatype=XSD.date)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@prefix dcat: <http://www.w3.org/ns/dcat#> .\n",
      "@prefix dcterms: <http://purl.org/dc/terms/> .\n",
      "@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .\n",
      "@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .\n",
      "\n",
      "<https://g-c8521c.81de.36fe.data.globus.org/Bayraktar/dt_Bayraktar_ClinicalMetadata_mock_2022.07.21.xlsx> a dcat:Dataset ;\n",
      "    rdfs:label \"dt_Bayraktar_ClinicalMetadata_mock_2022.07.21.xlsx\"^^xsd:string ;\n",
      "    dcterms:format <http://www.wikidata.org/entity/Q63082925> ;\n",
      "    dcterms:issued \"2022-08-23\"^^xsd:date ;\n",
      "    dcterms:modified \"2022-08-22\"^^xsd:date .\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(g.serialize(format=\"turtle\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>MPBTP (Sequencing Data)</th>\n",
       "      <th># of blocks</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age (optional)</th>\n",
       "      <th>Tumour Spatial Information</th>\n",
       "      <th>Pathology - pre-surgical</th>\n",
       "      <th>Histopathological Assignment</th>\n",
       "      <th>Histopathology - Microscopic Description</th>\n",
       "      <th>Mutations</th>\n",
       "      <th>Methylation Status (&gt;10% = METHYLATED)</th>\n",
       "      <th>Other Variants</th>\n",
       "      <th>Covid test</th>\n",
       "      <th>Other comments</th>\n",
       "      <th>Spatial Location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Patient A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>M</td>\n",
       "      <td>54</td>\n",
       "      <td>Right Temporal Lobe</td>\n",
       "      <td>GBM</td>\n",
       "      <td>Glioblastoma, IDH-wiltype (CNS WHO Grade 4)</td>\n",
       "      <td>Cellular glial neoplasm with elongated nuclei ...</td>\n",
       "      <td>IDH-WT (IHC and Sequencing)\\nATRX - Retained</td>\n",
       "      <td>MGMT Promotor Methylation - 2% - ABSENT</td>\n",
       "      <td>EGFR Amplification;\\nCDKN2A Deletion;\\nPTEN De...</td>\n",
       "      <td>Negative</td>\n",
       "      <td>A1 has a duplicate block</td>\n",
       "      <td>Wasn’t able to get stealth location informatio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Patient B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>F</td>\n",
       "      <td>82</td>\n",
       "      <td>Right Temporal Lobe</td>\n",
       "      <td>GBM</td>\n",
       "      <td>Glioblastoma, IDH-wiltype (CNS WHO Grade 4)</td>\n",
       "      <td>Pleomorphic astrocytes with numerous gemistocy...</td>\n",
       "      <td>IDH-WT (IHC and Sequencing)\\nATRX - Retained</td>\n",
       "      <td>MGMT Promotor Methylation - 5% - ABSENT</td>\n",
       "      <td>CDKN2A Deletion; \\nPTEN Deletion</td>\n",
       "      <td>Negative</td>\n",
       "      <td>-</td>\n",
       "      <td>Anterior, Lateral, Posterior and Medial sample...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Patient C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>F</td>\n",
       "      <td>41</td>\n",
       "      <td>Right Frontal Lobe</td>\n",
       "      <td>GBM</td>\n",
       "      <td>Oligodendroglioma, IDH-mut and 1p/19q co-delet...</td>\n",
       "      <td>Diffusely infiltrative glial neoplasm with mod...</td>\n",
       "      <td>IDH1-mut R132H (IHC and Sequencing)\\nATRX - Re...</td>\n",
       "      <td>No tested</td>\n",
       "      <td>1p/19q Co-deletion; \\nTP53 P152L</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Sampled by Richard</td>\n",
       "      <td>Superior, Anterior, Deep and Inferior samples ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Sample ID  MPBTP (Sequencing Data)  # of blocks Sex  Age (optional)  \\\n",
       "0  Patient A                      NaN            5   M              54   \n",
       "1  Patient B                      NaN            4   F              82   \n",
       "2  Patient C                      NaN            4   F              41   \n",
       "\n",
       "  Tumour Spatial Information  Pathology - pre-surgical  \\\n",
       "0         Right Temporal Lobe                     GBM    \n",
       "1         Right Temporal Lobe                     GBM    \n",
       "2          Right Frontal Lobe                     GBM    \n",
       "\n",
       "                        Histopathological Assignment  \\\n",
       "0        Glioblastoma, IDH-wiltype (CNS WHO Grade 4)   \n",
       "1        Glioblastoma, IDH-wiltype (CNS WHO Grade 4)   \n",
       "2  Oligodendroglioma, IDH-mut and 1p/19q co-delet...   \n",
       "\n",
       "            Histopathology - Microscopic Description  \\\n",
       "0  Cellular glial neoplasm with elongated nuclei ...   \n",
       "1  Pleomorphic astrocytes with numerous gemistocy...   \n",
       "2  Diffusely infiltrative glial neoplasm with mod...   \n",
       "\n",
       "                                           Mutations  \\\n",
       "0       IDH-WT (IHC and Sequencing)\\nATRX - Retained   \n",
       "1       IDH-WT (IHC and Sequencing)\\nATRX - Retained   \n",
       "2  IDH1-mut R132H (IHC and Sequencing)\\nATRX - Re...   \n",
       "\n",
       "    Methylation Status (>10% = METHYLATED)  \\\n",
       "0  MGMT Promotor Methylation - 2% - ABSENT   \n",
       "1  MGMT Promotor Methylation - 5% - ABSENT   \n",
       "2                                No tested   \n",
       "\n",
       "                                      Other Variants Covid test  \\\n",
       "0  EGFR Amplification;\\nCDKN2A Deletion;\\nPTEN De...   Negative   \n",
       "1                   CDKN2A Deletion; \\nPTEN Deletion   Negative   \n",
       "2                   1p/19q Co-deletion; \\nTP53 P152L   Negative   \n",
       "\n",
       "             Other comments                                   Spatial Location  \n",
       "0  A1 has a duplicate block  Wasn’t able to get stealth location informatio...  \n",
       "1                         -  Anterior, Lateral, Posterior and Medial sample...  \n",
       "2        Sampled by Richard  Superior, Anterior, Deep and Inferior samples ...  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.set_option('display.max_columns', None)\n",
    "xls = pd.ExcelFile(filename)\n",
    "subjects = pd.read_excel(xls)\n",
    "subjects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "BASE = Namespace(\"http://idr.semscape.org/\")\n",
    "SUBJECT = Namespace(BASE.subject)\n",
    "OBO = Namespace(\"http://purl.obolibrary.org/obo/\")\n",
    "g.bind(\"obo\", \"http://purl.obolibrary.org/obo/\")\n",
    "SCHEMA = Namespace(\"https://schema.org/\")\n",
    "SIO = Namespace(\"http://semanticscience.org/resource/\")\n",
    "NODE = Namespace(\"http://idr.semscape.org/nodeURI/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'Subject'",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mKeyError\u001B[0m                                  Traceback (most recent call last)",
      "File \u001B[0;32m~/projects/scheduled-bots/venv/lib/python3.8/site-packages/pandas/core/indexes/base.py:3621\u001B[0m, in \u001B[0;36mIndex.get_loc\u001B[0;34m(self, key, method, tolerance)\u001B[0m\n\u001B[1;32m   3620\u001B[0m \u001B[38;5;28;01mtry\u001B[39;00m:\n\u001B[0;32m-> 3621\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_engine\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mget_loc\u001B[49m\u001B[43m(\u001B[49m\u001B[43mcasted_key\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m   3622\u001B[0m \u001B[38;5;28;01mexcept\u001B[39;00m \u001B[38;5;167;01mKeyError\u001B[39;00m \u001B[38;5;28;01mas\u001B[39;00m err:\n",
      "File \u001B[0;32m~/projects/scheduled-bots/venv/lib/python3.8/site-packages/pandas/_libs/index.pyx:136\u001B[0m, in \u001B[0;36mpandas._libs.index.IndexEngine.get_loc\u001B[0;34m()\u001B[0m\n",
      "File \u001B[0;32m~/projects/scheduled-bots/venv/lib/python3.8/site-packages/pandas/_libs/index.pyx:163\u001B[0m, in \u001B[0;36mpandas._libs.index.IndexEngine.get_loc\u001B[0;34m()\u001B[0m\n",
      "File \u001B[0;32mpandas/_libs/hashtable_class_helper.pxi:5198\u001B[0m, in \u001B[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001B[0;34m()\u001B[0m\n",
      "File \u001B[0;32mpandas/_libs/hashtable_class_helper.pxi:5206\u001B[0m, in \u001B[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001B[0;34m()\u001B[0m\n",
      "\u001B[0;31mKeyError\u001B[0m: 'Subject'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001B[0;31mKeyError\u001B[0m                                  Traceback (most recent call last)",
      "Input \u001B[0;32mIn [7]\u001B[0m, in \u001B[0;36m<cell line: 1>\u001B[0;34m()\u001B[0m\n\u001B[1;32m      1\u001B[0m \u001B[38;5;28;01mfor\u001B[39;00m index, row \u001B[38;5;129;01min\u001B[39;00m subjects\u001B[38;5;241m.\u001B[39miterrows():\n\u001B[1;32m      2\u001B[0m     \u001B[38;5;66;03m# Row Subject\u001B[39;00m\n\u001B[0;32m----> 3\u001B[0m     g\u001B[38;5;241m.\u001B[39madd((SUBJECT[\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124m/\u001B[39m\u001B[38;5;124m\"\u001B[39m\u001B[38;5;241m+\u001B[39m\u001B[43mrow\u001B[49m\u001B[43m[\u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mSubject\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m]\u001B[49m], RDF\u001B[38;5;241m.\u001B[39mtype, OBO\u001B[38;5;241m.\u001B[39mNCIT_C16960))\n\u001B[1;32m      4\u001B[0m     g\u001B[38;5;241m.\u001B[39madd((SUBJECT[\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124m/\u001B[39m\u001B[38;5;124m\"\u001B[39m\u001B[38;5;241m+\u001B[39mrow[\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mSubject\u001B[39m\u001B[38;5;124m\"\u001B[39m]], DCTERMS\u001B[38;5;241m.\u001B[39midentifier, Literal(row[\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mSubject\u001B[39m\u001B[38;5;124m\"\u001B[39m], datatype\u001B[38;5;241m=\u001B[39mXSD\u001B[38;5;241m.\u001B[39mstring)))\n\u001B[1;32m      6\u001B[0m     subjectProperties \u001B[38;5;241m=\u001B[39m BNode()\n",
      "File \u001B[0;32m~/projects/scheduled-bots/venv/lib/python3.8/site-packages/pandas/core/series.py:958\u001B[0m, in \u001B[0;36mSeries.__getitem__\u001B[0;34m(self, key)\u001B[0m\n\u001B[1;32m    955\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_values[key]\n\u001B[1;32m    957\u001B[0m \u001B[38;5;28;01melif\u001B[39;00m key_is_scalar:\n\u001B[0;32m--> 958\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_get_value\u001B[49m\u001B[43m(\u001B[49m\u001B[43mkey\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m    960\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m is_hashable(key):\n\u001B[1;32m    961\u001B[0m     \u001B[38;5;66;03m# Otherwise index.get_value will raise InvalidIndexError\u001B[39;00m\n\u001B[1;32m    962\u001B[0m     \u001B[38;5;28;01mtry\u001B[39;00m:\n\u001B[1;32m    963\u001B[0m         \u001B[38;5;66;03m# For labels that don't resolve as scalars like tuples and frozensets\u001B[39;00m\n",
      "File \u001B[0;32m~/projects/scheduled-bots/venv/lib/python3.8/site-packages/pandas/core/series.py:1069\u001B[0m, in \u001B[0;36mSeries._get_value\u001B[0;34m(self, label, takeable)\u001B[0m\n\u001B[1;32m   1066\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_values[label]\n\u001B[1;32m   1068\u001B[0m \u001B[38;5;66;03m# Similar to Index.get_value, but we do not fall back to positional\u001B[39;00m\n\u001B[0;32m-> 1069\u001B[0m loc \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mindex\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mget_loc\u001B[49m\u001B[43m(\u001B[49m\u001B[43mlabel\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m   1070\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mindex\u001B[38;5;241m.\u001B[39m_get_values_for_loc(\u001B[38;5;28mself\u001B[39m, loc, label)\n",
      "File \u001B[0;32m~/projects/scheduled-bots/venv/lib/python3.8/site-packages/pandas/core/indexes/base.py:3623\u001B[0m, in \u001B[0;36mIndex.get_loc\u001B[0;34m(self, key, method, tolerance)\u001B[0m\n\u001B[1;32m   3621\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_engine\u001B[38;5;241m.\u001B[39mget_loc(casted_key)\n\u001B[1;32m   3622\u001B[0m \u001B[38;5;28;01mexcept\u001B[39;00m \u001B[38;5;167;01mKeyError\u001B[39;00m \u001B[38;5;28;01mas\u001B[39;00m err:\n\u001B[0;32m-> 3623\u001B[0m     \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;167;01mKeyError\u001B[39;00m(key) \u001B[38;5;28;01mfrom\u001B[39;00m \u001B[38;5;21;01merr\u001B[39;00m\n\u001B[1;32m   3624\u001B[0m \u001B[38;5;28;01mexcept\u001B[39;00m \u001B[38;5;167;01mTypeError\u001B[39;00m:\n\u001B[1;32m   3625\u001B[0m     \u001B[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001B[39;00m\n\u001B[1;32m   3626\u001B[0m     \u001B[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001B[39;00m\n\u001B[1;32m   3627\u001B[0m     \u001B[38;5;66;03m#  the TypeError.\u001B[39;00m\n\u001B[1;32m   3628\u001B[0m     \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_check_indexing_error(key)\n",
      "\u001B[0;31mKeyError\u001B[0m: 'Subject'"
     ]
    }
   ],
   "source": [
    "for index, row in subjects.iterrows():\n",
    "    # Row Subject\n",
    "    patientID = rllib.parse.quote(row[\"Sample ID\"])\n",
    "    g.add((SUBJECT[\"/\"+patientID], RDF.type, OBO.NCIT_C16960))\n",
    "    g.add((SUBJECT[\"/\"+patientID], DCTERMS.identifier, Literal(patientID, datatype=XSD.string)))\n",
    "    \n",
    "    subjectProperties = BNode()\n",
    "    g.add((SUBJECT[\"/\"+patientID], OBO.RO_0000053, subjectProperties))\n",
    "    # Row Age\n",
    "    ## please store birthdate, since Age does not age well in data. \n",
    "    \n",
    "    # g.add((subjectProperties, SCHEMA.birthDate, Literal(str(int(row[\"Age (year of birth)\"])), datatype=XSD.date)))\n",
    "    \n",
    "      \n",
    "\n",
    "    \n",
    "    # Family history\n",
    "    familyhistory = BNode()\n",
    "    g.add((familyhistory, RDF.type, RDF.Bag))\n",
    "    g.add((familyhistory, RDFS.domain, OBO.NCIT_C17726))\n",
    "    g.add((subjectProperties, OBO.BFO_0000185, familyhistory))\n",
    "    if row[\"Premen BRCA\"] == 1:\n",
    "        premenbrca = BNode()\n",
    "        g.add((familyhistory, RDFS.member, premenbrca))\n",
    "        g.add((premenbrca, RDF.type, OBO.NCIT_C49152))\n",
    "        g.add((premenbrca, RDFS.domain, OBO.DOID_1612))\n",
    "        g.add((premenbrca, URIRef(\"http://rs.tdwg.org/dwc/terms/lifeStage\"), OBO.OMIT_0018003))\n",
    "        g.add((premenbrca, RDF.value, Literal(int(row[\"Premen BRCA\"]), datatype=XSD.int)))\n",
    "\n",
    "    if row[\"Post BRCA\"] == 1:\n",
    "        postbrca = BNode()\n",
    "        g.add((familyhistory, RDFS.member, postbrca))\n",
    "        g.add((postbrca, RDF.type, OBO.NCIT_C49152))\n",
    "        g.add((postbrca, RDFS.domain, OBO.DOID_1612))\n",
    "        g.add((postbrca, URIRef(\"http://rs.tdwg.org/dwc/terms/lifeStage\"), OBO.OMIT_0023947))\n",
    "        g.add((postbrca, RDF.value, Literal(int(row[\"Post BRCA\"]), datatype=XSD.int)))\n",
    "\n",
    "    if row[\"OVCA\"] == 1:\n",
    "        ovca = BNode()\n",
    "        g.add((familyhistory, RDFS.member, ovca))\n",
    "        g.add((ovca, RDF.type, OBO.NCIT_C49152))\n",
    "        g.add((ovca, RDFS.domain, OBO.DOID_2394))\n",
    "        g.add((ovca, RDF.value, Literal(int(row[\"OVCA\"]), datatype=XSD.int)))\n",
    "        \n",
    "    if row[\"Male BrCA\"] == 1:\n",
    "        malebrca = BNode()\n",
    "        g.add((familyhistory, RDFS.member, malebrca))\n",
    "        g.add((malebrca, RDF.type, OBO.NCIT_C15709))\n",
    "        g.add((malebrca, RDFS.domain, OBO.DOID_1614))\n",
    "        g.add((malebrca, RDF.value, Literal(int(row[\"OVCA\"]), datatype=XSD.int)))\n",
    "        \n",
    "    \n",
    "    if row[\"BRCA1.1\"] == 1:\n",
    "        fambrca1test = BNode()\n",
    "        g.add((familyhistory, RDFS.member, fambrca1test))\n",
    "        g.add((fambrca1test, RDF.type, OBO.NCIT_C15709))\n",
    "        g.add((fambrca1test, RDFS.domain, OBO.NCIT_C17965))\n",
    "         \n",
    "    if row[\"BRCA2.1\"] == 1:\n",
    "        fambrca2test = BNode()\n",
    "        g.add((familyhistory, RDFS.member, fambrca2test))\n",
    "        g.add((fambrca1test, RDF.type, OBO.NCIT_C15709))\n",
    "        g.add((fambrca2test, RDFS.domain, OBO.NCIT_C18120))\n",
    "    if row[\"PALB2\"] == 1:\n",
    "        fampalb2test = BNode()\n",
    "        g.add((familyhistory, RDFS.member, fampalb2test))\n",
    "        g.add((fambrca1test, RDF.type, OBO.NCIT_C15709))\n",
    "        g.add((fampalb2test, RDFS.domain, OBO.NCIT_C84945))\n",
    "    if row[\"Other\"] == 1:\n",
    "        othertest = BNode()\n",
    "        g.add((familyhistory, RDFS.member, othertest))\n",
    "        g.add((othertest, RDF.type, OBO.NCIT_C15709))\n",
    "        g.add((fampalb2test, RDFS.domain, OBO.NCIT_C113067))\n",
    "   \n",
    "    # Prior Diagnosis\n",
    "    medicalhistory = BNode()\n",
    "    g.add((medicalhistory, RDF.type, RDF.Bag))\n",
    "    g.add((medicalhistory, RDFS.domain, OBO.SIO_010673))\n",
    "    g.add((subjectProperties, OBO.BFO_0000185, medicalhistory))\n",
    "    if row[\"Breast.1\"] == 1:\n",
    "        past_diagnosis = BNode()\n",
    "        g.add((medicalhistory, RDFS.member, past_diagnosis))\n",
    "        g.add((past_diagnosis, RDF.type, OBO.C4872))\n",
    "        g.add((past_diagnosis, DC.date, Literal(str(int(row[\"Year  Dx\"]))+\"-\"+str(int(row[\"Month Dx\"])), datatype=XSD.date)))\n",
    "      \n",
    "    # Menopause\n",
    "    ## premenopause\n",
    "    if row[\"Pre\"] == 1:\n",
    "        g.add((subjectProperties, URIRef(\"http://rs.tdwg.org/dwc/terms/lifeStage\"), OBO.OMIT_0018003))\n",
    "    if row[\"Peri\"] == 1:\n",
    "        g.add((subjectProperties, URIRef(\"http://rs.tdwg.org/dwc/terms/lifeStage\"), OBO.OMIT_0023947))\n",
    "    if row[\"Post\"] == 1:\n",
    "        g.add((subjectProperties, URIRef(\"http://rs.tdwg.org/dwc/terms/lifeStage\"), OBO.OMIT_0018004))\n",
    "\n",
    "    subjectArchivedSpecimen = BNode()\n",
    "    g.add((SUBJECT[\"/\"+row[\"Subject\"]], OBO.OBI_0100051, subjectArchivedSpecimen))\n",
    "    g.add((subjectArchivedSpecimen, RDF.type, OBO.NCIT_C19157))\n",
    "    \n",
    "    #Diagnosis\n",
    "    diagnosis = BNode()\n",
    "    g.add((subjectArchivedSpecimen, OBO.RO_0040035, diagnosis))\n",
    "    g.add((medicalhistory, RDFS.member, diagnosis))\n",
    "    if row[\"Invasive CA\"] == 1:\n",
    "        g.add((diagnosis, RDF.type, OBO.NCIT_C9480)) # Invasive carcinoma\n",
    "    if row[\"DCIS\"] == 1:\n",
    "        g.add((diagnosis, RDF.type, OBO.NCIT_C2924)) # DCIS\n",
    "    if row[\"LCIS\"] == 1:\n",
    "        g.add((diagnosis, RDF.type, OBO.NCIT_C137839)) # LCIS = Breast Pleomorphic Lobular Carcinoma In Situ?\n",
    "    if row[\"Atypia\"] == 1:\n",
    "        g.add((diagnosis, RDF.type, OBO.NCIT_C9478)) # Atypia\n",
    "    g.add((diagnosis, DC.date, Literal(str(int(row[\"Year\"]))+\"-\"+str(int(row[\"Month\"])), datatype=XSD.date)))\n",
    " \n",
    "    # How should \"Normal\" be interpreted?\n",
    "    \n",
    "    tfinding = BNode()\n",
    "    g.add((diagnosis, OBO.NCIT_R108, tfinding))\n",
    "    g.add((tfinding, RDF.type, OBO.NCIT_C48879))\n",
    "    g.add((tfinding, RDF.value, Literal(\"T\"+str(row[\"T\"]), datatype=XSD.string)))\n",
    "    \n",
    "    nfinding = BNode()\n",
    "    g.add((diagnosis, OBO.NCIT_R108, nfinding))\n",
    "    g.add((nfinding, RDF.type, OBO.NCIT_C48879))\n",
    "    g.add((nfinding, RDF.value, Literal(\"N\"+str(row[\"N\"]), datatype=XSD.string)))\n",
    "    \n",
    "    mfinding = BNode()\n",
    "    g.add((diagnosis,OBO.NCIT_R108, mfinding))\n",
    "    g.add((mfinding, RDF.type, OBO.NCIT_C48879))\n",
    "    g.add((mfinding, RDF.value, Literal(\"M\"+str(row[\"M\"]), datatype=XSD.string)))\n",
    "    \n",
    "    # Expression\n",
    "    if not pd.isna(row[\"ER\"]):\n",
    "        er = BNode()\n",
    "        g.add((subjectArchivedSpecimen, OBO.TXPO_0001873, er))\n",
    "        g.add((er, RDF.type, OBO.NCIT_C25209))\n",
    "        g.add((er, RDFS.range, OBO.NCIT_C16150))\n",
    "        if row[\"ER\"] == 1:\n",
    "            g.add((er, RDF.value, Literal(True)))\n",
    "        if row[\"ER\"] == 0:\n",
    "            g.add((er, RDF.value, Literal(False)))\n",
    "            \n",
    "    if not pd.isna(row[\"PR\"]):\n",
    "        pr = BNode()\n",
    "        g.add((subjectArchivedSpecimen, OBO.TXPO_0001873, pr))\n",
    "        g.add((pr, RDF.type, OBO.NCIT_C25209))\n",
    "        g.add((pr, RDFS.range, OBO.NCIT_C16149))\n",
    "        if row[\"PR\"] == 1:\n",
    "            g.add((pr, RDF.value, Literal(True)))\n",
    "        if row[\"PR\"] == 0:\n",
    "            g.add((pr, RDF.value, Literal(False)))\n",
    "            \n",
    "    if not pd.isna(row[\"HER2\"]):\n",
    "        her = BNode()\n",
    "        g.add((subjectArchivedSpecimen, OBO.TXPO_0001873, her))\n",
    "        g.add((her, RDF.type, OBO.NCIT_C25209))\n",
    "        g.add((her, RDFS.range, OBO.NCIT_C184942))\n",
    "        if row[\"HER2\"] == 1:\n",
    "            g.add((her, RDF.value, Literal(True)))\n",
    "        if row[\"HER2\"] == 0:\n",
    "            g.add((her, RDF.value, Literal(False)))\n",
    "    \n",
    "    if not pd.isna(row[\"Ki67 (%)\"]):\n",
    "        ki67 = BNode()\n",
    "        g.add((subjectArchivedSpecimen, OBO.TXPO_0001873, ki67))\n",
    "        g.add((her, RDF.type, OBO.NCIT_C25209))\n",
    "        g.add((her, RDF.value, Literal(row[\"Ki67 (%)\"])))\n",
    "        g.add((her, SIO.SIO_000221, SIO.SIO_001413))\n",
    "        \n",
    "            \n",
    "    # collection method\n",
    "    collectionmethod = BNode()\n",
    "    g.add((subjectArchivedSpecimen, OBO.HSO_0000288, collectionmethod))\n",
    "    g.add((collectionmethod, RDF.type, OBO.NCIT_C70700))\n",
    "    if row[\"core Biopsy (y/n)\"] == 1:\n",
    "        g.add((collectionmethod, RDF.value, OBO.NCIT_C15189))\n",
    "    if row[\"Excisional biopsy\"] == 1:\n",
    "        g.add((collectionmethod, RDF.value, OBO.NCIT_C51633))\n",
    "    if row[\"Mastectomy\"] == 1:\n",
    "        g.add((collectionmethod, RDF.value, OBO.NCIT_C15279))\n",
    "    \n",
    "    # Hypoxia time (mins)\n",
    "    specimenHandling = BNode()\n",
    "    g.add((subjectArchivedSpecimen, OBO.RO_0002334, specimenHandling))\n",
    "    g.add((specimenHandling, RDF.type, OBO.NCIT_C179745))\n",
    "    g.add((specimenHandling, RDFS.range, OBO.OMIT_0016357))\n",
    "    g.add((specimenHandling, RDF.value, Literal(row[\"Hypoxia time (mins)\"])))\n",
    "    g.add((specimenHandling, SIO.SIO_000221, SIO.SIO_000434)) # unit minutes \n",
    "    \n",
    "    # Formalin Fix Time (mins)\n",
    "    specimenHandling = BNode()\n",
    "    g.add((subjectArchivedSpecimen, OBO.RO_0002334, specimenHandling))\n",
    "    g.add((specimenHandling, RDF.type, OBO.NCIT_C179745))\n",
    "    g.add((specimenHandling, RDFS.range, OBO.OBIB_0000718))\n",
    "    g.add((specimenHandling, RDF.value, Literal(row[\"Formalin Fix Time\"])))\n",
    "    g.add((specimenHandling, SIO.SIO_000221, SIO.SIO_000434)) # unit minutes \n",
    "    \n",
    "    # Mutations\n",
    "    if row[\"BRCA1\"] == 1:\n",
    "        known_mutation = BNode()\n",
    "        g.add((SUBJECT[\"/\"+row[\"Subject\"]], OBO.CLO_0037333, known_mutation))\n",
    "        g.add((known_mutation, RDF.type, OBO.NCIT_C36391))\n",
    "        g.add((known_mutation, RDFS.range, OBO.NCIT_C17965))\n",
    "        g.add((known_mutation, RDFS.comment, Literal(row[\"Specific MT (text)\"])))\n",
    "    if row[\"BRCA2\"] == 1:\n",
    "        known_mutation = BNode()\n",
    "        g.add((SUBJECT[\"/\"+row[\"Subject\"]], OBO.CLO_0037333, known_mutation))\n",
    "        g.add((known_mutation, RDF.type, OBO.NCIT_C36391))\n",
    "        g.add((known_mutation, RDFS.range, OBO.NCIT_C18120))\n",
    "        g.add((known_mutation, RDFS.comment, Literal(row[\"Specific MT (text)\"])))\n",
    "    if row[\"other\"] == 1:\n",
    "        known_mutation = BNode()\n",
    "        g.add((SUBJECT[\"/\"+row[\"Subject\"]], OBO.CLO_0037333, known_mutation))\n",
    "        g.add((known_mutation, RDF.type, OBO.NCIT_C36391))\n",
    "        g.add((known_mutation, RDFS.range, OBO.NCIT_C16612))\n",
    "        g.add((known_mutation, RDFS.comment, Literal(row[\"Specific MT (text)\"])))\n",
    "    \n",
    "    # Treatment\n",
    "    if row[\"NACT (y/n)\"] == 1:\n",
    "        treatment = BNode()\n",
    "        g.add((subjectArchivedSpecimen, WDT.P2176, treatment))\n",
    "        g.add((treatment, RDF.type, OBO.NCIT_C15665))\n",
    "        if not pd.isna(row[\"Year.1\"]):\n",
    "            g.add((treatment, SCHEMA.startDate, Literal(str(row[\"Year.1\"])+\"-\"+str(row[\"Month.1\"]), datatype=XSD.date)))\n",
    "        if not pd.isna(row[\"Year.2\"]):\n",
    "            g.add((treatment, SCHEMA.endDate, Literal(str(row[\"Year.2\"])+\"-\"+str(row[\"Month.2\"]), datatype=XSD.date)))\n",
    "    if row[\"PARP-I y/n\"] == 1:\n",
    "        treatment = BNode()\n",
    "        g.add((subjectArchivedSpecimen, WDT.P2176, treatment))\n",
    "        g.add((treatment, RDF.type, OBO.NCIT_C62554))\n",
    "        if not pd.isna(row[\"Year.1\"]):\n",
    "            g.add((treatment, SCHEMA.startDate, Literal(str(row[\"Year.1\"])+\"-\"+str(row[\"Month.1\"]), datatype=XSD.date)))\n",
    "        if not pd.isna(row[\"Year.2\"]):\n",
    "            g.add((treatment, SCHEMA.endDate, Literal(str(row[\"Year.2\"])+\"-\"+str(row[\"Month.2\"]), datatype=XSD.date)))\n",
    "   \n",
    "    \n",
    "        \n",
    "        \n",
    "    \n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(g.serialize(destination=filename+\".ttl\", format=\"turtle\"))\n",
    "print(g.serialize(format=\"turtle\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for node in g.all_nodes():\n",
    "    if isinstance(node, URIRef):\n",
    "        try:\n",
    "            g.parse(node)\n",
    "        except:\n",
    "            print(node)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g.parse(\"http://www.wikidata.org/entity/Q63082925.ttl\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extract schema"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from shexer.shaper import Shaper\n",
    "from shexer.consts import NT, TURTLE\n",
    "\n",
    "q = \"select ?class where { ?item rdf:type ?class }\"\n",
    "target_classes = []\n",
    "x = g.query(q)\n",
    "for target_class in x:\n",
    "  if str(target_class[\"class\"]) not in target_classes:\n",
    "    target_classes.append(str(target_class[\"class\"]))\n",
    "\n",
    "shex_target_file = filename+\".shex\"\n",
    "\n",
    "shaper = Shaper(target_classes=target_classes,\n",
    "                rdflib_graph=g,\n",
    "                input_format=TURTLE,\n",
    "                )  # Default rdf:type\n",
    "            \n",
    "shaper.shex_graph(output_file=shex_target_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "command = f\"cat {filename}.shex\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.system(command)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}